### ISSUE TITLE
Make empty  search_around_* return arrays have reasonable units

### ISSUE THREAD

**Issue by eteq** [astropy/astropy#5518]
This is a follow-on prompted by @adrn's https://github.com/astropy/astropy/pull/5083#r69383542 in  #5083. The key question is why, when `search_around_*` yield nothing, are the units of the resulting empty vectors deg and dimensionless unscaled instead of having the dimensions of the `SkyCoord`s that they're called on?  I think the answer is simply that that seemed most straightforward to implement, but probably the better thing to do is to give back whatever unit you would have gotten if the arrays had not been empty.

Note that it would be best to make this consistent across `search_around_*` *and* `match_coordinates_*`, although the latter hasn't yet gotten the treatment from #5083 to make it yield anything in the case of empty coordinates, so this might be easier to do piecemeal.

I don't think I'll have time, but if @evertrol or @adrn might want to take a shot at this, we could try to slip it in by the feature freeze in 5 days (it's a backwards-incompatible change so can't go in a bugfix)

**Comment by evertrol** [comment]
I prefer the option to return a dimension, so that potentially a user or script can proceed in the same manner as it would if the result were non-empty (it may crash later, but then astropy isn't to blame).
The angular unit appears to be set to that of the first coordinate list, or degree if that is empty, so presumably something similar goes for the distance unit.
In fact, an empty input list should still have a unit attached, so probably both the `d2ds` and `d3ds` should be set to the relevant units from `coords1`, and not a standard of `degree` either (as with `dimensionless_unscaled`, I copy-pasted that for consistency and simplicity). Though it appears that `SkyCoord` automatically normalizes its units to degree anyway.
The only time where that doesn't work, would be in `search_around_sky` when `coords1` does not have distance information. As is currently, it simply falls back to `dimensionless_unscaled`, which seems to be the right approach with empty results and input as well.

In summary, my preference would be:
- always have `d2ds` and `d3ds` use the units of the `coord1` input argument.
- sole exception would be `d3ds` in `search_around_sky` when there is no distance information: use `dimensionless_unscaled` instead.

**Comment by evertrol** [comment]
As for the `match_coordinates_*` functions: I thought about it and had given it a try, but it only breaks when the `catalog` argument is empty: if the `matchcoord` argument is empty, it returns a an empty `Angle` list. I felt that for the empty catalog case, it'd be on the head of the user to have it break, although the resulting error message is not directly clear (`ValueError: zero-size array to reduction operation maximum which has no identity`, which bubbles up from scipy's `KDTree`).
I see a few options:
- leave at is
- raise an exception with a clearer error message (`"catalog is empty"`)
- return empty matches, as for `search_around_*`.

**Comment by evertrol** [comment]
Minor addendum: I came across a discrepancy between `SkyCoord.separation_3d` and `SkyCoord.search_around_3d` in this context: the former return value has a distance unit equal to the `SkyCoord` instance, the latter return value has a distance unit equal to its argument.
The obvious cause is the order of the arguments as they are passed through to the respective `separation_3d` and `search_around_3d` functions, with `self` being the first and second argument, respectively.

This is obviously a minor thing, but with the above discussion which units the return value should take, it irks me, as it seems inconsistent. Not sure if this is worth bothering with (there may be a good reason for the ordering).

**Comment by eteq** [comment]
Good point re `match_to_catalog_*`, @evertrol - I like your suggestion of a slightly more meaningful exception.  Will make a quick PR to do that, which should then close this issue.

**Comment by evertrol** [comment]
Thanks @eteq 

Is it worth filing an issue for the order in the `search_around_` functions?
Going by the [documentation for `SkyCoord.search_around_sky`](http://docs.astropy.org/en/stable/api/astropy.coordinates.SkyCoord.html#astropy.coordinates.SkyCoord.search_around_sky), I'm inclined to say it's a bug even, as the order of the returned indices is contrary to what is stated in the documentation. Unless I'm misunderstanding something there.

**Comment by eteq** [comment]
@evertrol - oh, right, sorry, forgot about that comment above.

To address why it's that way: as I recall it's due to an implementation detail of the `matching.search_around_*` functions: the *second* argument is the one that doesn't get transformed, which means it keeps its kD tree.  So if you do ``something.search_around_?(something_else)`` and then ``something.search_around_?(something_else2)``, you get a speedup from not having to create the kdtree, but only for the second argument.

That said, it sounds like the `search_around_sky` docs are wrong, so definitely a bug there.  Furthermore, it's not really clear that the above interpretation is any better than the alternative (that the *second* argument keep the kD tree) and I see your point that it's more consistent with the other functions.  So I'm definitely not averse to a PR fixing that (and given that it's nominally a bug, it's not subject to the feature freeze restrictions).

**Comment by evertrol** [comment]
@eteq Thanks for the kdtree explanation.
After starting a bit more at it, I realised it's indeed mainly a documentation bug, but only for the *description* part of the first two return values of the two relevant `SkyCoord` methods; the names of the return values appear to be actually correct, even if not used (and the various examples elsewhere in the docs also show the correct usage).

I'll put in a PR to get that fixed for 1.3 (and, presumably, 1.0.9).

Is it worth opening an issue for reversing the order of the returned indices? This is quite a dangerous change (even if for astropy 2.0), since a lot of existing code could break (or worse, returning incorrect results without people noticing), so it may need some discussion whether this should be done at all.

**Comment by eteq** [comment]
@evertrol - thanks - sounds good to do a PR for the docs, and then separately an issue on actually changing the order (which would be for 2.0).. Just be aware that I will be resistant to that for just the reason you said about the "danger" - but you can try to convince me 😉
